File: lld/problems/trading-system/go/tradingsystem.go
// Package tradingsystem implements a simplified order matching engine LLD
// problem: a per-symbol order book that matches buy/sell limit orders under
// price-time priority, supports partial fills and cancellation.
package tradingsystem

import (
	"errors"
	"fmt"
	"sync"
	"time"
)

type Side int

const (
	Buy Side = iota
	Sell
)

type OrderType int

const (
	Limit OrderType = iota
	Market
)

type OrderStatus int

const (
	Open OrderStatus = iota
	PartiallyFilled
	Filled
	Cancelled
)

type Order struct {
	ID            string
	Symbol        string
	Side          Side
	Type          OrderType
	Price         float64
	Quantity      int
	RemainingQty  int
	Timestamp     time.Time
	Status        OrderStatus
	seq           int64
}

type Trade struct {
	ID          string
	Symbol      string
	BuyOrderID  string
	SellOrderID string
	Price       float64
	Quantity    int
	Timestamp   time.Time
}

var ErrOrderNotFound = errors.New("order not found or already filled/cancelled")

// OrderBook holds resting orders for a single symbol, sorted by price-time
// priority: buyOrders descending by price then ascending by seq (arrival
// order), sellOrders ascending by price then ascending by seq.
type OrderBook struct {
	Symbol string

	mu         sync.Mutex
	buyOrders  []*Order
	sellOrders []*Order
	orders     map[string]*Order
	trades     []Trade
	seq        int64
	tradeSeq   int64
}

func NewOrderBook(symbol string) *OrderBook {
	return &OrderBook{
		Symbol: symbol,
		orders: make(map[string]*Order),
	}
}

// SubmitOrder matches the incoming order against the resting opposite side
// while prices cross, returns the resulting trades, and rests any
// unfilled remainder in the book (LIMIT orders only; MARKET orders never
// rest — an unfilled remainder is simply cancelled).
//
// The whole book is locked for the duration of matching: concurrent
// submissions to the same book must be serialized, otherwise two
// goroutines could both match against the same resting order and
// double-fill it.
func (ob *OrderBook) SubmitOrder(order *Order) []Trade {
	ob.mu.Lock()
	defer ob.mu.Unlock()

	ob.seq++
	order.seq = ob.seq
	order.RemainingQty = order.Quantity
	order.Status = Open
	ob.orders[order.ID] = order

	var trades []Trade
	if order.Side == Buy {
		trades = ob.matchIncoming(order, &ob.sellOrders, func(restingPrice float64) bool {
			return order.Type == Market || order.Price >= restingPrice
		})
	} else {
		trades = ob.matchIncoming(order, &ob.buyOrders, func(restingPrice float64) bool {
			return order.Type == Market || order.Price <= restingPrice
		})
	}

	if order.RemainingQty == 0 {
		order.Status = Filled
		delete(ob.orders, order.ID)
		return trades
	}

	if order.Type == Market {
		order.Status = Cancelled
		delete(ob.orders, order.ID)
		return trades
	}

	if len(trades) > 0 {
		order.Status = PartiallyFilled
	}
	if order.Side == Buy {
		ob.buyOrders = insertSorted(ob.buyOrders, order, buyLess)
	} else {
		ob.sellOrders = insertSorted(ob.sellOrders, order, sellLess)
	}
	return trades
}

// matchIncoming repeatedly crosses order against the front of restingSide
// while crosses(price) holds, producing trades and shrinking/removing
// resting orders as they get filled.
func (ob *OrderBook) matchIncoming(order *Order, restingSide *[]*Order, crosses func(restingPrice float64) bool) []Trade {
	var trades []Trade
	for order.RemainingQty > 0 && len(*restingSide) > 0 {
		resting := (*restingSide)[0]
		if !crosses(resting.Price) {
			break
		}

		qty := order.RemainingQty
		if resting.RemainingQty < qty {
			qty = resting.RemainingQty
		}

		ob.tradeSeq++
		trade := Trade{
			Symbol:    ob.Symbol,
			Price:     resting.Price,
			Quantity:  qty,
			Timestamp: time.Now(),
		}
		if order.Side == Buy {
			trade.BuyOrderID, trade.SellOrderID = order.ID, resting.ID
		} else {
			trade.BuyOrderID, trade.SellOrderID = resting.ID, order.ID
		}
		trade.ID = fmt.Sprintf("TR-%d", ob.tradeSeq)
		trades = append(trades, trade)
		ob.trades = append(ob.trades, trade)

		order.RemainingQty -= qty
		resting.RemainingQty -= qty

		if resting.RemainingQty == 0 {
			resting.Status = Filled
			delete(ob.orders, resting.ID)
			*restingSide = (*restingSide)[1:]
		} else {
			resting.Status = PartiallyFilled
		}
	}
	return trades
}

// CancelOrder removes a still-resting order from the book. It fails if the
// order does not exist or has already been filled/cancelled.
func (ob *OrderBook) CancelOrder(orderID string) error {
	ob.mu.Lock()
	defer ob.mu.Unlock()

	order, ok := ob.orders[orderID]
	if !ok {
		return ErrOrderNotFound
	}
	order.Status = Cancelled
	delete(ob.orders, orderID)

	if order.Side == Buy {
		ob.buyOrders = removeByID(ob.buyOrders, orderID)
	} else {
		ob.sellOrders = removeByID(ob.sellOrders, orderID)
	}
	return nil
}

// BuyOrders and SellOrders return snapshots of the resting book, best
// priority first, for inspection/testing.
func (ob *OrderBook) BuyOrders() []*Order {
	ob.mu.Lock()
	defer ob.mu.Unlock()
	out := make([]*Order, len(ob.buyOrders))
	copy(out, ob.buyOrders)
	return out
}

func (ob *OrderBook) SellOrders() []*Order {
	ob.mu.Lock()
	defer ob.mu.Unlock()
	out := make([]*Order, len(ob.sellOrders))
	copy(out, ob.sellOrders)
	return out
}

func (ob *OrderBook) Trades() []Trade {
	ob.mu.Lock()
	defer ob.mu.Unlock()
	out := make([]Trade, len(ob.trades))
	copy(out, ob.trades)
	return out
}

func buyLess(a, b *Order) bool {
	if a.Price != b.Price {
		return a.Price > b.Price
	}
	return a.seq < b.seq
}

func sellLess(a, b *Order) bool {
	if a.Price != b.Price {
		return a.Price < b.Price
	}
	return a.seq < b.seq
}

// insertSorted inserts o into the already-sorted slice, keeping the
// less(a,b) ordering. Linear scan is fine at interview scope; a real
// exchange would index price levels with a balanced tree or heap.
func insertSorted(orders []*Order, o *Order, less func(a, b *Order) bool) []*Order {
	i := 0
	for i < len(orders) && less(orders[i], o) {
		i++
	}
	orders = append(orders, nil)
	copy(orders[i+1:], orders[i:])
	orders[i] = o
	return orders
}

func removeByID(orders []*Order, id string) []*Order {
	for i, o := range orders {
		if o.ID == id {
			return append(orders[:i], orders[i+1:]...)
		}
	}
	return orders
}

File: lld/problems/trading-system/python/trading_system.py
"""Trading system LLD — Python reference implementation.

A per-symbol order book that matches buy/sell limit (and market) orders
under price-time priority, supports partial fills and cancellation. See
../go/tradingsystem.go for the original design writeup.
"""
from __future__ import annotations

import itertools
import threading
import time
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Callable, List


class Side(Enum):
    BUY = auto()
    SELL = auto()


class OrderType(Enum):
    LIMIT = auto()
    MARKET = auto()


class OrderStatus(Enum):
    OPEN = auto()
    PARTIALLY_FILLED = auto()
    FILLED = auto()
    CANCELLED = auto()


class OrderNotFoundError(Exception):
    """Raised when cancelling an order that doesn't exist or already
    filled/cancelled."""

    def __init__(self) -> None:
        super().__init__("order not found or already filled/cancelled")


@dataclass
class Order:
    id: str
    symbol: str
    side: Side
    type: OrderType
    price: float
    quantity: int
    remaining_qty: int = field(init=False)
    timestamp: float = field(default_factory=time.time)
    status: OrderStatus = field(default=OrderStatus.OPEN)
    seq: int = field(default=0, repr=False)

    def __post_init__(self) -> None:
        self.remaining_qty = self.quantity


@dataclass
class Trade:
    id: str
    symbol: str
    buy_order_id: str
    sell_order_id: str
    price: float
    quantity: int
    timestamp: float = field(default_factory=time.time)


class OrderBook:
    """Holds resting orders for a single symbol, sorted by price-time
    priority: buy orders descending by price then ascending by arrival
    sequence, sell orders ascending by price then ascending by arrival
    sequence.

    The whole book is locked for the duration of matching: concurrent
    submissions to the same book must be serialized, otherwise two threads
    could both match against the same resting order and double-fill it.
    """

    def __init__(self, symbol: str):
        self.symbol = symbol
        self._lock = threading.Lock()
        self._buy_orders: List[Order] = []
        self._sell_orders: List[Order] = []
        self._orders: dict[str, Order] = {}
        self._trades: List[Trade] = []
        self._seq = itertools.count(1)
        self._trade_seq = itertools.count(1)

    def submit_order(self, order: Order) -> List[Trade]:
        """Matches the incoming order against the resting opposite side
        while prices cross, returns the resulting trades, and rests any
        unfilled remainder in the book (LIMIT orders only; MARKET orders
        never rest — an unfilled remainder is simply cancelled)."""
        with self._lock:
            order.seq = next(self._seq)
            order.remaining_qty = order.quantity
            order.status = OrderStatus.OPEN
            self._orders[order.id] = order

            if order.side == Side.BUY:
                trades = self._match_incoming(
                    order,
                    self._sell_orders,
                    lambda resting_price: order.type == OrderType.MARKET or order.price >= resting_price,
                )
            else:
                trades = self._match_incoming(
                    order,
                    self._buy_orders,
                    lambda resting_price: order.type == OrderType.MARKET or order.price <= resting_price,
                )

            if order.remaining_qty == 0:
                order.status = OrderStatus.FILLED
                self._orders.pop(order.id, None)
                return trades

            if order.type == OrderType.MARKET:
                order.status = OrderStatus.CANCELLED
                self._orders.pop(order.id, None)
                return trades

            if trades:
                order.status = OrderStatus.PARTIALLY_FILLED

            if order.side == Side.BUY:
                self._insert_sorted(self._buy_orders, order, buy=True)
            else:
                self._insert_sorted(self._sell_orders, order, buy=False)
            return trades

    def _match_incoming(
        self,
        order: Order,
        resting_side: List[Order],
        crosses: Callable[[float], bool],
    ) -> List[Trade]:
        """Repeatedly crosses order against the front of resting_side while
        crosses(price) holds, producing trades and shrinking/removing
        resting orders as they get filled."""
        trades: List[Trade] = []
        while order.remaining_qty > 0 and resting_side:
            resting = resting_side[0]
            if not crosses(resting.price):
                break

            qty = min(order.remaining_qty, resting.remaining_qty)

            if order.side == Side.BUY:
                buy_order_id, sell_order_id = order.id, resting.id
            else:
                buy_order_id, sell_order_id = resting.id, order.id

            trade = Trade(
                id=f"TR-{next(self._trade_seq)}",
                symbol=self.symbol,
                buy_order_id=buy_order_id,
                sell_order_id=sell_order_id,
                price=resting.price,
                quantity=qty,
            )
            trades.append(trade)
            self._trades.append(trade)

            order.remaining_qty -= qty
            resting.remaining_qty -= qty

            if resting.remaining_qty == 0:
                resting.status = OrderStatus.FILLED
                self._orders.pop(resting.id, None)
                resting_side.pop(0)
            else:
                resting.status = OrderStatus.PARTIALLY_FILLED

        return trades

    def cancel_order(self, order_id: str) -> None:
        """Removes a still-resting order from the book. Raises
        OrderNotFoundError if the order does not exist or has already been
        filled/cancelled."""
        with self._lock:
            order = self._orders.get(order_id)
            if order is None:
                raise OrderNotFoundError()
            order.status = OrderStatus.CANCELLED
            del self._orders[order_id]

            side = self._buy_orders if order.side == Side.BUY else self._sell_orders
            for i, o in enumerate(side):
                if o.id == order_id:
                    del side[i]
                    break

    def buy_orders(self) -> List[Order]:
        """Snapshot of resting buy orders, best priority first."""
        with self._lock:
            return list(self._buy_orders)

    def sell_orders(self) -> List[Order]:
        """Snapshot of resting sell orders, best priority first."""
        with self._lock:
            return list(self._sell_orders)

    def trades(self) -> List[Trade]:
        with self._lock:
            return list(self._trades)

    @staticmethod
    def _insert_sorted(orders: List[Order], order: Order, buy: bool) -> None:
        # Buy side: best price first = highest price, then earliest seq.
        # Sell side: best price first = lowest price, then earliest seq.
        price_key = -order.price if buy else order.price
        i = 0
        while i < len(orders):
            existing_key = -orders[i].price if buy else orders[i].price
            if (existing_key, orders[i].seq) <= (price_key, order.seq):
                i += 1
            else:
                break
        orders.insert(i, order)


def _demo() -> None:
    book = OrderBook("AAPL")
    book.submit_order(Order("S1", "AAPL", Side.SELL, OrderType.LIMIT, 100, 10))
    trades = book.submit_order(Order("B1", "AAPL", Side.BUY, OrderType.LIMIT, 100, 10))

    print("Trades:", trades)
    print("Resting buys:", book.buy_orders())
    print("Resting sells:", book.sell_orders())
    print("All trades:", book.trades())


if __name__ == "__main__":
    _demo()
